clients/src/main/java/org/apache/kafka/clients/producer/internals/DefaultPartitioner.java

```diff
@@ -68,6 +69,15 @@ public int partition(String topic, Object key, byte[] keyBytes, Object value, by
         }
     }
 
+    private int getNextValue(String topic) {
+        AtomicInteger counter = topicCounterMap.get(topic);
+        if (null == counter) {
+            counter = new AtomicInteger(new Random().nextInt());
+            topicCounterMap.put(topic, counter);
```

Comment:
`getNextValue` is not synchronised. There is a race in this line. Should be replaced with `counter = topicCounterMap.putIfAbsent(topic, counter);`